### PR TITLE
7.1: Support for new multidraw WebGL extensions

### PR DESCRIPTION
**Pull request by ibgreen**
# For #972

Note: This PR is intended as base for discussion, no urgency to land

#### Background
- Exploring base API support for multidraw extensions
#### Change List
- polyfill system
- feature system
- Program.draw(...)

### CODE REVIEW

File: modules/core/src/webgl/classes/program.js [hunk 1]
```diff
@@ -98,37 +98,25 @@ export default class Program extends Resource {
 
   // A good thing about the WebGL API is that there are so many ways to draw things ;)
   // This function unifies those ways into a single call using common parameters with sane defaults
-  draw({
```
Comment (ibgreen) [inline modules/core/src/webgl/classes/program.js]: draw function is getting really big, splitting it in pieces, maybe should move to separate file

File: modules/core/src/webgl/classes/program.js [hunk 2]
```diff
@@ -214,6 +195,136 @@ export default class Program extends Resource {
...
+
+    const {isIndexed, isInstanced, drawMode, indexType} = options;
+
+    const multiDrawExt = this.gl.getExtension('WEBGL_multi_draw');
+    const multiDrawInstancedExt = this.gl.getExtension('WEBGL_multi_draw_instanced');
+    // TODO - In WebGL2 we could fallback to drawRangeElements for non instanced rendering?
+    assert(multiDrawExt && multiDrawInstancedExt);
+
+    if (isIndexed && isInstanced) {
+      multiDrawInstancedExt.multiDrawElementsInstancedWEBGL(
```
Comment (ibgreen) [inline modules/core/src/webgl/classes/program.js]: Unfortunately it is quite a verbose API, hopefully will minify reasonably so the bundle size impact of supporting it stays small.

Should do before/after measurements...

File: modules/core/src/webgl/features/webgl-features-table.js [hunk 3]
```diff
@@ -31,6 +34,9 @@ export const FEATURES = {
   COLOR_ATTACHMENT_FLOAT: 'COLOR_ATTACHMENT_FLOAT',
   COLOR_ATTACHMENT_HALF_FLOAT: 'COLOR_ATTACHMENT_HALF_FLOAT',
 
+  // POST WEBGL2 EXTENSIONS
```
Comment (ibgreen) [inline modules/core/src/webgl/features/webgl-features-table.js]: Whoops, will remove

File: modules/core/src/webgl/features/webgl-features-table.js [hunk 4]
```diff
@@ -69,6 +80,11 @@ export default {
   [FEATURES.COLOR_ATTACHMENT_FLOAT]: [false, 'EXT_color_buffer_float'],
   [FEATURES.COLOR_ATTACHMENT_HALF_FLOAT]: [false, 'EXT_color_buffer_half_float'],
 
+  // EXPERIMENTAL EXTENSIONS
+  [FEATURES.MULTIDRAW]: [
```
Comment (ibgreen) [inline modules/core/src/webgl/features/webgl-features-table.js]: Ditto

File: modules/webgl2-polyfill/src/polyfill-context.js [hunk 5]
```diff
@@ -10,7 +10,11 @@
 /* eslint-disable camelcase, brace-style */
 import assert from './assert';
 
-import {WEBGL2_CONTEXT_POLYFILLS, WEBGL2_CONTEXT_OVERRIDES} from './polyfill-table';
+import {
+  WEBGL2_CONTEXT_POLYFILLS,
+  // WEBGL2PLUS_CONTEXT_POLYFILLS,
```
Comment (ibgreen) [inline modules/webgl2-polyfill/src/polyfill-context.js]: Should we support WebGL2+ polyfills here? This is pretty clearly a "WebGL3" (OpenGLES 3.x) API...

File: modules/webgl2-polyfill/src/polyfills/multidraw-polyfill.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
+/* eslint-disable max-params */
+
+export function multiDrawArraysPolyfill(
```
Comment (ibgreen) [inline modules/webgl2-polyfill/src/polyfills/multidraw-polyfill.js]: as far as I can see, these extensions can be polyfilled client side (without performance benefits of course)...

TBD - is this the right thing to do?

**Comment by tsherif** [comment]
Stongly recommend building this around an example that uses both of the new functions. Plus have to think about how programs should be set up to change uniforms.

**Comment by ibgreen** [comment]
> Strongly recommend building this around an example that uses both of the new functions. 

Agreed, makes no sense to land this without at least a demo. A demo that just calls the functions would be trivial to put together, one that really demonstrates the benefits would be more work.

> Plus have to think about how programs should be set up to change uniforms.

Yes, this extension does not allow different uniforms in the same draw call, it just allows rendering pieces of a geometry in a single call.

The way I have used something like this in the past is that I had a function that merged geometries with the same material into a larger geometry.

In this case, since we can specify ranges, we can even merge geometries with different materials into a single geometry and a set of ranges for each material...

Of course that line of thought is very scenegraph oriented. In the general sense uniforms do not come from just materials, so the system may need to be more generic...

**Comment by tsherif** [comment]
> Yes, this extension does not allow different uniforms in the same draw call, it just allows rendering pieces of a geometry in a single call.

My understanding was that the power of this was in being able to switch uniforms based on the gl_DrawID variable available in the shaders. Otherwise, I don't really see what the win is...?

**Comment by tsherif** [comment]
I.e. like this: https://github.com/kenrussell/webgl-animometer/blob/master/Animometer/tests/3d/webgl.html#L64-L68

I think that usage is the key value proposition of these extensions.

**Comment by ibgreen** [comment]
> My understanding was that the power of this was in being able to switch uniforms based on the gl_DrawID variable available in the shaders. 

Ah, that would be awesome! How would that work? Having uniform arrays indexed by the draw id?

```
The gl_DrawID builtin is added to the shading language. For any Multi* instanced draw call variant, the index of the draw i may be read by the vertex shader as gl_DrawID. For non Multi* calls, the value of gl_DrawID is 0.
```

> Otherwise, I don't really see what the win is...?

I have used geometry merging techniques in the past to get 500+ draw calls down to 5, made a ~2x difference on FPS. But it does require merging geometries.

This extension gives more flexibility in that regard, but obviously a lot less useful than if we can also handle different uniforms in the same call.

**Comment by tsherif** [comment]
> I have used geometry merging techniques in the past to get 500+ draw calls down to 5, made a ~2x difference on FPS. But it does require merging geometries.

Ah right. At that scale it makes sense you'd see a win in CPU overhead. I think you'd still need to merge geo, right, since it all has to be a the same buffers?

> Ah, that would be awesome! How would that work? Having uniform arrays indexed by the draw id?

That's what I imagined. Does Luma currently support uniform arrays?

**Comment by ibgreen** [comment]
Yep that makes sense. Uniform buffers are probably something we want to adopt/support more widely before or in parallel with this feature. Sounds like a good 7.1 feature combo?

**Pull request closed without merging**